docs/the-new-architecture/pillars-fabric-components.md

```diff
@@ -26,7 +28,84 @@ To create a Fabric Component, we have to follow these steps:
 
-### Javascript Specification
+### Folder Setup
+
+The easiest way to create a Component is as a separate module we will then import as a dependency for our apps. This has several benefits, among which it pushes for code reuse, it keeps the components decoupled from the rest of the apps and it makes it easier to leverage some automation already in place.
+
+For this guide, we are going to create a Fabric Component that centers some text on the screen.
+
+Let's create a new folder at the same level of our app and let's call it `RTNCenteredText`.
+
+In this folder, we are going to create three subfolders: `js`, `ios` and `android`.
+
+The final result should look like this:
+
+<figure>
+  <img width="500" alt="Folder Structure for a Fabric Component" src="/docs/assets/NewArchitecture/AppFolderStructure.png"/>
+  <figcaption>Initial folder structure for a Fabric Component.</figcaption>
+</figure>
+
+### JavaScript Specification
+
+In the **New Architecture**, we decided to have a single source of truth for Components API and we decided that it has to be a typed dialect of JavaScript (either [Flow](https://flow.org/) or [TypeScript](https://www.typescriptlang.org/)). We need a typed dialect because the **Codegen** has to generate some code in C++, Objective-C++ and Java which are all strongly typed languages: therefore, it needs information on types.
+
+Another important aspect of the JavaScript specification is the file name. A Component filename has to end with the `NativeComponent.js` (or `jsx`, `ts`, `tsx`) suffix. The **Codegen** process will look up for files ending with that suffix to generate the required code.
+
+The following are the specification of our `RTNCenteredText` Component in both Flow and TypeScript.
+
+<Tabs groupId="fabric-component-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
+<TabItem value="flow">
+
+```typescript
+// @flow strict-local
+
+import type {ViewProps} from 'react-native/Libraries/Components/View/ViewPropTypes';
+import type {HostComponent} from 'react-native';
+import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';
+
+type NativeProps = $ReadOnly<{|
+  ...ViewProps,
+  text: ?string,
+  // add other props here
+|}>;
+
+export default (codegenNativeComponent<NativeProps>(
+   'RTNCenteredText',
+): HostComponent<NativeProps>);
+```
+
+</TabItem>
+<TabItem value="typescript">
+
+```typescript
+import type { ViewProps } from 'ViewPropTypes';
+import type { HostComponent } from 'react-native';
+import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';
+
+export interface NativeProps extends ViewProps {
+  ...ViewProps,
+  text: ?string,
+  // add other props here
+}
+
+export default codegenNativeComponent<NativeProps>(
+  'RTNCenteredText'
+) as HostComponent<NativeProps>;
```

Comment:
I'm not a typescript expert. I had followed something that has already been delivered in the official documentation [here](https://reactnative.dev/docs/next/new-architecture-library-intro). I will stick to this for consistency: we may have a project to update the typescript types and the associated docs in the next future. ;)

Reply:
Gotcha, maybe just a good thing to start a discussion to track in the New Architecture working group then? I have to catch up on what is there

Reply:
> I'm not a typescript expert. I had followed something that has already been delivered in the official documentation [here](https://reactnative.dev/docs/next/new-architecture-library-intro)

Yeah, this isn't exactly wrong, but it's an escape hatch and not the ideal solution :) we're saying "it's definitely this type, don't correct me" to the compiler, which is of course less ideal than actually typing the output of `codegenNativeComponent<T>` to be `HostComponent<T>`, since we could be wrong

Reply:
Oh, I see your point. 
Could it be that it has been done in this way because we don't have a TS type for the `codegenNativeComponent<T>` function yet?

Reply:
> Could it be that it has been done in this way because we don't have a TS type for the codegenNativeComponent<T> function yet?

Yeah, I believe that's correct from what I see. Is that something the definitely-typed community repo would eventually provide?